executor/hash_table.go

```diff
@@ -43,12 +45,60 @@ const (
 	estCountDivisor = 8
 )
 
+type bitmap struct {
+	bits []byte
+}
+
+func (bm *bitmap) clear() {
+	for i := 0; i < len(bm.bits); i++ {
+		bm.bits[i] = byte(0)
+	}
+}
+
+func (bm *bitmap) reserve(size int) {
+	size = (size + 7) / 8
+	if len(bm.bits) < size {
+		bm.bits = make([]byte, size)
+	} else {
+		bm.clear()
+	}
+}
+
+func (bm *bitmap) isSet(idx int) bool {
+	b := bm.bits[idx/8] & (1 << uint(idx%8))
+	return b != 0
+}
+
 // hashContext keeps the needed hash context of a db table in hash join.
 type hashContext struct {
 	allTypes  []*types.FieldType
 	keyColIdx []int
-	h         hash.Hash64
-	buf       []byte
+
+	// hashVals stores the encoded value of each chunk row.
+	// The real element type of hashVals must be `hash.Hash64`.
+	hashVals   []io.Writer
+	buf        []byte
+	nullBitmap bitmap
```

Comment:
It could be more effective and simple to just use `[]bool` instead of a bitmap because the memory usage of `hashContext ` here is not that important.

Reply:
have done